src/compiler/tracing.ts

```diff
@@ -92,40 +92,38 @@ namespace ts.tracing {
         Emit = "emit",
     }
 
-    export type EventData = [phase: Phase, name: string, args?: object];
-
-    /** Note: `push`/`pop` should be used by default.
-     * `begin`/`end` are for special cases where we need the data point even if the event never
-     * terminates (typically for reducing a scenario too big to trace to one that can be completed).
-     * In the future we might implement an exit handler to dump unfinished events which would
-     * deprecate these operations.
-     */
-    export function begin(phase: Phase, name: string, args?: object) {
-        if (!traceFd) return;
-        writeEvent("B", phase, name, args);
-    }
-    export function end(phase: Phase, name: string, args?: object) {
-        if (!traceFd) return;
-        writeEvent("E", phase, name, args);
-    }
-
     export function instant(phase: Phase, name: string, args?: object) {
         if (!traceFd) return;
         writeEvent("I", phase, name, args, `"s":"g"`);
     }
 
     // Used for "Complete" (ph:"X") events
-    const completeEvents: { phase: Phase, name: string, args?: object, time: number }[] = [];
-    export function push(phase: Phase, name: string, args?: object) {
+    const completeEvents: { phase: Phase, name: string, args?: object, time: number, separateBeginAndEnd: boolean }[] = [];
+
+    /**
+     * @param separateBeginAndEnd - used for special cases where we need the trace point even if the event
+     * never terminates (typically for reducing a scenario too big to trace to one that can be completed).
+     * In the future we might implement an exit handler to dump unfinished events which would deprecate
+     * these operations.
+     */
+    export function push(phase: Phase, name: string, args?: object, separateBeginAndEnd = false) {

```

Comment:
Ooh, so the main thing in all of this is the `separateBeginAndEnd`?  Good way to switch early to a push/pop!

Reply:
Sorry, I should have called that out more explicilty.  Yes, `begin` -> `push(/*separateBeginAndEnd*/ true)`.